SCAutolib/models/gui.py

```diff
@@ -0,0 +1,304 @@
+        grayscale = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+        upscaled = cv2.resize(grayscale,
+                              dsize=None,
+                              fx=UPSCALING_FACTOR,
+                              fy=UPSCALING_FACTOR,
+                              interpolation=cv2.INTER_LANCZOS4)
+        _, binary = cv2.threshold(upscaled, 120, 255, cv2.THRESH_BINARY_INV)
+        image_data_str = pytesseract.image_to_data(binary)
+        df = pd.read_csv(StringIO(image_data_str),
+                         sep='\t', lineterminator='\n')
+
+        yres, xres = binary.shape[:2]
+        df[['left', 'width']] /= xres
+        df[['top', 'height']] /= yres
+
+        logger.debug(df)
+        return df
+
+    @staticmethod
+    def _images_same(filename1: str, filename2: str):
+        """Compare two images, return True if they are completely identical."""
+        im1 = cv2.imread(filename1)
+        im2 = cv2.imread(filename2)
+
+        # Is the resolution the same
+        if im1.shape != im2.shape:
+            return False
+
+        # Check if value of every pixel is the same
+        if np.bitwise_xor(im1, im2).any():
+            return False
+
+        return True
+
+    def click_on(self, key: str, timeout: float = 30):
+        """Clicks matching word on the screen.
+
+        """
+        logger.info(f"Trying to find key='{key}' to click on.")
+
+        end_time = time() + timeout
+        item = None
+        first_scr = None
+
+        # Repeat screenshotting, until the key is found
+        while time() < end_time:
+            # Capture the screenshot
+            screenshot = self.screen.screenshot()
+
+            last_scr = screenshot
+            if first_scr is None:
+                first_scr = screenshot
+
+            df = self._image_to_data(screenshot)
+            selection = df['text'] == key
+
+            # If there is no matching word, try again
+            if selection.sum() == 0:
+                logger.info('Found no match, trying again')
+                continue
+
+            # Exactly one word matching, exit the loop
+            elif selection.sum() == 1:
+                logger.info('Found exactly one match')
+                item = df[selection]
+                break
+
+            # More than one word matches, choose the first match
+            # Probably deterministic, but it should not be relied upon
+            else:
+                logger.info('Found multiple matches')
+                item = df.iloc[0]
+                break
+
+        if item is None:
+            raise Exception(f"Found no key='{key}' in screenshots " +
+                            f"{first_scr} to {last_scr}")
+
+        x = float(item['left'] + item['width']/2)
+        y = float(item['top'] + item['height']/2)
+
+        self.mouse.move(x, y)
+        sleep(0.5)
+        self.mouse.click()
+        sleep(self.WAIT_TIME)
+
+    def assert_text(self, key: str, timeout: float = 0):
```

Comment:
This method could be (after some modifications) used in "click on" as the code is redundant right now.

Reply:
I agree that some part of the code is redundant. However, I think that creating a new method for doing this would make the code more complicated in the end.